waddrmgr/scoped_manager.go

```diff
@@ -818,33 +818,46 @@ func (s *ScopedKeyManager) nextAddresses(ns walletdb.ReadWriteBucket,
 		}
 	}
 
-	// Finally update the next address tracking and add the addresses to
-	// the cache after the newly generated addresses have been successfully
-	// added to the db.
 	managedAddresses := make([]ManagedAddress, 0, len(addressInfo))
 	for _, info := range addressInfo {
 		ma := info.managedAddr
-		s.addrs[addrKey(ma.Address().ScriptAddress())] = ma
-
-		// Add the new managed address to the list of addresses that
-		// need their private keys derived when the address manager is
-		// next unlocked.
-		if s.rootManager.IsLocked() && !s.rootManager.WatchOnly() {
-			s.deriveOnUnlock = append(s.deriveOnUnlock, info)
-		}
-
 		managedAddresses = append(managedAddresses, ma)
 	}
 
-	// Set the last address and next address for tracking.
-	ma := addressInfo[len(addressInfo)-1].managedAddr
-	if internal {
-		acctInfo.nextInternalIndex = nextIndex
-		acctInfo.lastInternalAddr = ma
-	} else {
-		acctInfo.nextExternalIndex = nextIndex
-		acctInfo.lastExternalAddr = ma
+	// Finally, create a closure that will update the next address tracking
+	// and add the addresses to the cache after the newly generated
+	// addresses have been successfully committed to the db.
+	onCommit := func() {
+		// Since this closure will be called when the DB transaction
```

Comment:
I'm not so sure about this. [The `OnCommit` callback is called from _within_ the `Commit` method](https://github.com/boltdb/bolt/blob/master/tx.go#L230), as a result, we'll still have the mutex held [since it's grabbed here](https://github.com/btcsuite/btcwallet/blob/7dc4cdcc9af3c50ce9918137878b9d69a3e43c43/waddrmgr/scoped_manager.go#L1062).

Reply:
This is tricky, but I believe the order of operations will be
```
db tx opened
    mutex locked
    mutex unlocked
    onCommit called
           mutex locked
           mutex unlocked
    onCommit done
db tx committed
```

Note that the mutex is unlocked before `NextExternalAddresses` returns, while the db tx gets commited after it has returned.

Reply:
Yes, it's unlocked before `NextExternalAddresses` returns, but it'll only return _after_ `nextAddresses` returns. Instead, I think it's more like: 
```
mutex locked
db tx opened
    onCommit called
           mutex locked
           mutex unlocked
    onCommit done
db tx committed
mutex unlocked
```

Notice the outer locking of the mutex _outside_ the database transaction.

Reply:
But `ns walletdb.ReadWriteBucket` is passed _into_ `NextExternalAddresses`, which means that the db tx gets opened **before** the mutex is locked?

Reply:
Ahh, yeah you're right! Thanks for walking through this, time to land this ting.

Reply:
I think there's still an issue with usage of `Batch`, but we don't do so atm, so it shouldn't pop up. Worst case would be address re-use AFAICT.